### PR TITLE
Show node output type right under the node, not in pop-up.

### PR DESCRIPTION
**Pull request by s9ferech**
### Pull Request Description

This solves issue #1401.

(outdated screenshot)
Screenshot 2021-04-02 091401

The type appears on hover: (outdated recording)
main

It does not appear when a visualization is visible or we have an error: (outdated recordings)
visualization
error

Types are abbreviated for a short moment before the full appears:
abbreviation

### Checklist
Please include the following checklist in your PR:

- [x] The `CHANGELOG.md` was updated with the changes introduced in this PR.
- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.
- [x] All code has automatic tests where possible.
- [x] All code has been profiled where possible.
- [x] All code has been manually tested in the IDE.
- [x] All code has been manually tested in the "debug/interface" scene.
- [x] All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).
- [ ] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

### CODE REVIEW

File: src/rust/ide/view/graph-editor/src/component/node.rs
Comment (s9ferech) [inline src/rust/ide/view/graph-editor/src/component/node.rs]: This part seems problematic. But I don't know how we initialize networks properly.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (s9ferech) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: This part duplicates some logic from the shape definitions, but it seems difficult to avoid that.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (s9ferech) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs:516]: This is quite redundant. Can this be avoided?

**Comment by s9ferech** [comment]
Currently, the labels clash with outgoing wires. How should that be solved?

File: CHANGELOG.md
Comment (farmaazon) [inline CHANGELOG.md]: I suppose the description should be altered, as we already released the pop-ups.

File: src/rust/ide/view/graph-editor/src/component/node.rs
Comment (farmaazon) [inline src/rust/ide/view/graph-editor/src/component/node.rs]: I think I included such line in one of my commits.

File: src/rust/ide/view/graph-editor/src/component/node.rs
Comment (farmaazon) [reply src/rust/ide/view/graph-editor/src/component/node.rs]: Well, the root cause is that each frp node has output being an output value until some input events arrives.

Knowing that, let's look at the line: 
```
            model.output.set_port_label_visibility
                <+ visualization_visible.not().and(&no_error_set);
```
the visualization_visible is properly false, however no_error_set could not - until someone emits signal about new error. I'm not sure why disabling/enabling visualizations is necessary - @wdanilo do you have any thoughts?

If we want to have port types displayed by default, we should directly fire model.output.set_port_label_visibility with true during initialization

File: CHANGELOG.md
Comment (wdanilo) [inline CHANGELOG.md]: This information should be a nice sounding sentence to the user. It shoudl inform him about a new feature in the product. Please work with Ara on formulating it in this spirit.

File: src/rust/ide/view/graph-editor/src/component/node.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node.rs]: What this line has to do with "showing node output type under the node" (which is the subject of this PR)?

File: src/rust/ide/view/graph-editor/src/component/node.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node.rs]: Few things here:

1. Do not leave such comments in the code, please. This is a PR, not a Draft PR. It means that you propose these changes. So you propose that we add to code a comment "All three are needed in this order, for reasons that I do not understand.". Until this issue is resolved, this should be marked as a Draft PR, not a real PR.
2. `frp.enable_visualization();frp.disable_visualization();` looks bad. Take a look at Adam's explanation above and design the FRP netwrok initialization based on what Adam has described.

File: src/rust/ide/view/graph-editor/src/component/node/output/area.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/area.rs]: What port labels have to do with types under the node? This change seems completely unrelated to the subject of this PR.

File: src/rust/ide/view/graph-editor/src/component/node/output/area.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/area.rs]: What port labels have to do with types under the node? This change seems completely unrelated to the subject of this PR.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: What port labels have to do with types under the node? This change seems completely unrelated to the subject of this PR.

**EDIT:** Ah, these are not port labels. They were meant to be output type labels! The naming have to be changed eveywhere to reflect that.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: This is port model. So in port model we have a field label. Is this label of a port?

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs:443]: Do you really need to store this information?

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [reply src/rust/ide/view/graph-editor/src/component/node/output/port.rs:516]: If its redundant, why not just delete it?

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [reply src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: Please talk with Michael, who was implementing the existing equations if reusing them could be possible. Maybe we should not use `Var<f32>` in the equations but just `f32`? I mean - this might be possible, depending on usage. Please confirm this with Michael - for me, it seems like refactoring this logic should be doable, as we are really computing the same things there.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: I think you're looking for this
`label.set_content <+ frp.tp.map(|tp| tp.map_or_else(||default(),|s| s.to_string()))`

Comment (wdanilo) [reply src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: Btw, please add a function `map_or_default` to `Option` to our Prelude, and then use it like 
`label.set_content <+ frp.tp.map(|tp| tp.map_or_defaukt(|s| s.to_string()))`

This will be insanely useful function.

**Review by wdanilo — CHANGES_REQUESTED** [review]
Also, when creating such visual PRs, please show the result in action - create a screen recording includign different corner casees - moving your maouse with different zoom levels, connecting the nodes, dragging the connection over these texts, etc. We need to see all possible corner cases to understand if this behaves as it should!

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (s9ferech) [reply src/rust/ide/view/graph-editor/src/component/node/output/port.rs:516]: I think I only tried to be consistent with other code and I find it slightly weird to use something called `set_size` in places where I only read the size.

File: src/rust/ide/view/graph-editor/src/component/node.rs
Comment (s9ferech) [reply src/rust/ide/view/graph-editor/src/component/node.rs]: Firing `model.output.set_port_label_visibility` might not change anything since it would get overwritten by the network connection that you mentioned. I'll have to find the roots from where the incorrect values propagate.

Generally, I find it a bit problematic that I have to care about the consistent initialization of networks. The connections that I create already describe how it should be initialized. Is there a better way to do this?

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (s9ferech) [reply src/rust/ide/view/graph-editor/src/component/node/output/port.rs:443]: We need it to compute the position of the label whenever the width of the node changes. I think wee need to store it.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (s9ferech) [reply src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: I have created a [PR for the prelude](https://github.com/enso-org/rust-lib/pull/35). Who should review that?

**Comment by s9ferech** [comment]
There is still some code left in `node` and `graph-editor` that can be used to display tooltips. Will we keep that or should it be deleted? @farmaazon @wdanilo

**Comment by wdanilo** [comment]
Ok, so here's the visual review:

1. Something is wrong with spacing. On some screenshots, the text is very close to nodes:
    image
    while on some it is far away:
    image
2. Let's move the type information closer to the node, so when two nodes are created by the searcher, the type will appear between them. Please make the offset configurable in the theme editor.
3. Please make also the offset at which the node searcher makes new nodes configurable in the theme editor (if its not done yet).
4. Let's simplify the view - after hovering over the output port, lets display only the last part of the type, like `Vector`. Let's display the full type after keeping the mouse cursor there for a moment (approx 2-3 seconds - configurable in theme editor).

After having that done, I'll try to adjust the themes values and then I'd love everyone interested in this topic to play with it and have a conversation about this solution vs the popups one (surprisingly, this is the most controversial / hard design part so far).

File: src/rust/ide/view/graph-editor/src/component/node.rs
Comment (farmaazon) [reply src/rust/ide/view/graph-editor/src/component/node.rs]: Well, as I know, the FRP node output value is not overwritten be setting up network connection _per se_, but only when the event go through it. So we need to intialize the output value to have proper value before any input connection will be fired.

**Comment by s9ferech** [comment]
> 4. Let's simplify the view - after hovering over the output port, lets display only the last part of the type, like `Vector`. Let's display the full type after keeping the mouse cursor there for a moment (approx 2-3 seconds - configurable in theme editor).

Currently, this seems to be easy since we seem to only display single identifiers there. But will it stay like that? I assume that we will display full types, composed from multiple identifiers and operators, in the future.

**Comment by wdanilo** [comment]
What is the status of this task ?

**Comment by s9ferech** [comment]
> What is the status of this task ?

I made all the changes that you requested. It is ready for review. There is one new recording that shows the abbreviated types and the new label offset in action.

File: src/rust/ide/view/src/project.rs
Comment (wdanilo) [inline src/rust/ide/view/src/project.rs]: This is not using the StyleWatchFRP, so it would not update when the theme changes. Is this correct or am I missing something here? If its correct, using the frp version would be pretty simple, as we are close to frp network here.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: where should not be indented

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: spacing - space after `:`

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: please add docs to these functions - they are not obvious. For example, what is "border" here? And what is border length? In the next function, what is "segment" - these should be described in docs.

**Review by wdanilo — COMMENTED** [review]
This looks super ocol. The implementation looks good as well to me! Just a couple of comments. Also, could you not remove the popup and make it configurable - either using popup or tdisplayin it below the node? This is tricky as hell to get it right here, and Id prefer to have both options for now. Regarding the popup - please keep the abbreviated thing there as well, as the abbreviation is amazing!

File: src/rust/ide/view/src/project.rs
Comment (s9ferech) [reply src/rust/ide/view/src/project.rs]: I fixed it.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: This is not a sentence, missing dot. The same applies to a lit of other lines in other docs. Please fix this in all new docs.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [reply src/rust/ide/view/graph-editor/src/component/node/output/port.rs:516]: Felix, please do not mark such conversations as "resolved", as it was not resolved. You've replied, marked it as "resolved" and it was hidden. I'm reopenning it.

I still don't understand it. What do you mean by "This is quite redundant" then? I do not understand your last comment - could you elaborate, please? Let's not leave in code anything that is redundant. This needs to be fixed before merge.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: What are these variables used for? The names `relative_x` and `center_x` do not explain what they are used for. What is `relative_x` relative to? Could you please improve the names so their meaning would be obvious?

Also, the logic is strange IMO. Why are you adding and removing `0.5`? For example, what if `port_index` is `1` and `port_count` is `2`? Then your equation is `(1 + 0.5) / 2 - 0.5 = 1.5 / 2.0 - 0.5 = 0.25` - TBH I don't understand reasoning behind it. This should be explained much better.

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs:497]: What are you creating animation for if you are setting its duration to 0 ?

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/node/output/port.rs]: Please add comments here describing what these constants are used for (explaining that we have left 2 implementations for now).

**Review by wdanilo — APPROVED** [review]

File: src/rust/ide/view/graph-editor/src/component/node/output/port.rs
Comment (s9ferech) [reply src/rust/ide/view/graph-editor/src/component/node/output/port.rs:497]: I use it just for the delay. When the delay is over, I want to switch the value instantly. I copied this from how the way how visualizations are shown on hover. Is there a better way to do this?

**Review by farmaazon — APPROVED** [review]

**Pull request merged**
